puffin/src/scope_details.rs

```diff
@@ -0,0 +1,261 @@
+    pub(crate) string_to_scope_id: std::collections::HashMap<String, ScopeId>,
+}
+
+/// Provides fast read access to scope details.
+/// This collection can be cloned safely.
+#[derive(Default, Clone)]
+pub struct ScopeCollection(Arc<RwLock<Inner>>);
+
+impl ScopeCollection {
+    /// Provides read to the given closure for some scope details.
+    pub fn read_by_id<F: FnMut(&ScopeDetails)>(&self, scope_id: &ScopeId, mut cb: F) {
+        if let Some(read) = self.0.read().scope_id_to_details.get(scope_id) {
+            cb(read);
+        }
+    }
+
+    /// Provides read to the given closure for some scope details.
+    pub fn read_by_name<F: FnMut(&ScopeId)>(&self, scope_name: &str, mut cb: F) {
+        if let Some(read) = self.0.read().string_to_scope_id.get(scope_name) {
+            cb(read);
+        }
+    }
+
+    /// Only puffin should be allowed to allocate and provide the scope id so this function is private to puffin.
+    pub(crate) fn insert(&self, scope_details: ScopeDetails) {
+        assert!(scope_details.scope_id.is_some());
+
+        let id = scope_details.identifier();
+        self.0
+            .write()
+            .string_to_scope_id
+            .insert(id.to_string(), scope_details.scope_id.unwrap());
+        self.0.write().scope_id_to_details.insert(
+            scope_details.scope_id.unwrap(),
+            scope_details.into_readable(),
+        );
+    }
+
+    /// Manually register scope details. After a scope is inserted it can be reported to puffin.
+    pub fn register_custom_scopes(&self, scopes: &[ScopeDetails]) -> HashSet<ScopeId> {
+        let mut new_scopes = HashSet::new();
+        for scope_detail in scopes {
+            let new_scope_id = fetch_add_scope_id();
+            self.insert(scope_detail.clone().with_scope_id(new_scope_id));
+            new_scopes.insert(new_scope_id);
+        }
+
+        new_scopes
+    }
+
+    /// Fetches all registered scopes and their ids.
+    /// Useful for fetching scope id by a static scope name.
+    pub fn scopes_by_name<T>(
+        &self,
+        mut existing_scopes: impl FnMut(&std::collections::HashMap<String, ScopeId>) -> T,
+    ) -> T {
+        existing_scopes(&self.0.read().string_to_scope_id)
+    }
+
+    /// Fetches all registered scopes.
+    /// Useful for fetching scope details by a scope id.
+    pub fn scopes_by_id<T>(
+        &self,
+        mut existing_scopes: impl FnMut(&std::collections::HashMap<ScopeId, ScopeDetails>) -> T,
+    ) -> T {
+        existing_scopes(&self.0.read().scope_id_to_details)
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Hash, PartialOrd, Ord, Eq)]
+/// This struct contains scope details and can be read by external libraries.
+pub struct ScopeDetails {
+    /// Unique scope identifier.
+    // Always initialized once registered.
+    // It is `None` when an external library has yet to register this scope.
+    pub(crate) scope_id: Option<ScopeId>,
```

Comment:
If we use `ScopeId(pub NonZeroU32)` then this field will shrink in size, thanks to `enum niche optimization`